src/datasets/vision/imagenet.jl

```diff
@@ -0,0 +1,204 @@
+ 1
+ 1
+ 1
+ 1
+ 1
+
+julia> X, y = dataset[1:5];
+
+julia> size(X)
+(224, 224, 3, 5)
+
+julia> dataset.metadata
+Dict{String, Any} with 4 entries:
+  "class_WNIDs"       => ["n02119789", "n02100735", "n02110185", "n02096294", "n02102040", "n02066245", "n02509815", "n02124075", "n02417914", "n02123394"  …  "n02815834", "n09229709", "n07697313", "n03888605", "n03355925", "n03…
+  "class_description" => ["small grey fox of southwestern United States; may be a subspecies of Vulpes velox", "an English breed having a plumed tail and a soft silky coat that is chiefly white", "breed of sled dog developed in …
+  "class_names"       => Vector{SubString{String}}[["kit fox", "Vulpes macrotis"], ["English setter"], ["Siberian husky"], ["Australian terrier"], ["English springer", "English springer spaniel"], ["grey whale", "gray whale", "d…
+  "wnid_to_label"     => Dict("n07693725"=>768, "n03775546"=>829, "n01689811"=>469, "n02100877"=>192, "n02441942"=>48, "n04371774"=>569, "n07717410"=>741, "n03347037"=>919, "n04355338"=>526, "n02097474"=>158…)
+```
+"""
+struct ImageNet <: SupervisedDataset
+    metadata::Dict{String,Any}
+    split::Symbol
+    files::Vector{String}
+    targets::Vector{Int}
+    Tx::Type
+end
+
+ImageNet(; split=:train, Tx=Float32, dir=nothing) = ImageNet(Tx, split; dir)
+ImageNet(split::Symbol; kws...) = ImageNet(; split, kws...)
+ImageNet(Tx::Type; kws...) = ImageNet(; Tx, kws...)
+
+function ImageNet(
+    Tx::Type,
+    split::Symbol;
+    dir=nothing,
+    train_dir="train",
+    val_dir="val",
+    test_dir="test",
+    devkit_dir="devkit",
+)
+    @assert split ∈ (:train, :val, :test)
+
+    DEPNAME = "ImageNet"
+    METADATA_FILENAME = joinpath(devkit_dir, "data", "meta.mat")
+
+    TRAINSET_SIZE = 1_281_167
+    VALSET_SIZE = 50_000
+    TESTSET_SIZE = 100_000
+
+    # Load metadata
+    file_path = datafile(DEPNAME, METADATA_FILENAME, dir)
+    metadata = ImageNetReader.read_metadata(file_path)
+
+    root_dir = @datadep_str DEPNAME
+    if split == :train
+        files = ImageNetReader.readdata(joinpath(root_dir, train_dir))
+        @assert length(files) == TRAINSET_SIZE
+    elseif split == :val
+        files = ImageNetReader.readdata(joinpath(root_dir, val_dir))
+        @assert length(files) == VALSET_SIZE
+    else
+        files = ImageNetReader.readdata(joinpath(root_dir, test_dir))
+        @assert length(files) == TESTSET_SIZE
+    end
+    targets = [metadata["wnid_to_label"][wnid] for wnid in ImageNetReader.load_wnids(files)]
+    return ImageNet(metadata, split, files, targets, Tx)
+end
+
+function convert2image(::Type{<:ImageNet}, x::AbstractArray{<:Integer})
+    return convert2image(ImageNet, reinterpret(N0f8, convert(Array{UInt8}, x)))
+end
+convert2image(::Type{<:ImageNet}, x) = ImageNetReader.inverse_preprocess(x)
+
+Base.length(d::ImageNet) = length(d.image_files)
+function Base.getindex(d::ImageNet, ::Colon)
+    # Throw warning here that ImageNet probably will not fit in memory?
```

Comment:
If this almost certainly hits out-of-memory error, then we should probably just error eagerly, e.g., `Base.getindex(d::ImageNet, ::Colon) = error("Loading all images net images into memory is problematic.")`

Reply:
Added `ArgumentError` in 02e966d with a message to use `dataset[1:end]` if loading all of the dataset is indeed the desired behaviour.